### PR TITLE
Support deploying multiple injector replicas with auto-TLS

### PR DESCRIPTION
**Pull request by tomhjp**
This PR implements support for the leader election functionality added in https://github.com/hashicorp/vault-k8s/pull/198.

Acceptance test passing locally:

```bash
$ bats -t test/acceptance/injector-leader-elector.bats
1..1
ok 1 injector: testing leader elector
```

By default, no behaviour change other than some pod anti-affinity settings which I think are harmless, but please correct me if you think the default injector affinity settings should stay empty.

### CODE REVIEW

File: templates/injector-deployment.yaml [hunk 1]
```diff
@@ -84,6 +84,14 @@ spec:
             - name: AGENT_INJECT_TELEMETRY_PATH
               value: "/metrics"
             {{- end }}
+            {{- if and (eq (.Values.injector.leaderElector.enabled | toString) "true") (gt (.Values.injector.replicas | int) 1) }}
```
Comment (jasonodonnell) [inline templates/injector-deployment.yaml:87]: I see this being used a few times, should we just make it a helper function?

File: templates/injector-leader-endpoint.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{{- if and (eq (.Values.injector.enabled | toString) "true" ) (eq (.Values.global.enabled | toString) "true") (eq (.Values.injector.leaderElector.enabled | toString) "true") (gt (.Values.injector.replicas | int) 1) }}
+# This is created here so it can be cleaned up easily, since if
+# the endpoint is left around the leader won't expire for about a minute.
+apiVersion: v1
```
Comment (jasonodonnell) [inline templates/injector-leader-endpoint.yaml:4]: This needs Helm annotations:

```yaml
     app.kubernetes.io/name: {{ include "vault.name" . }}-agent-injector
     app.kubernetes.io/instance: {{ .Release.Name }}
     app.kubernetes.io/managed-by: {{ .Release.Service }}
```

File: values.yaml [hunk 3]
```diff
@@ -27,6 +27,14 @@ injector:
   # True if you want to enable vault agent injection.
   enabled: true
 
+  replicas: 1
+
+  # If multiple replicas are specified, by default a leader-elector side-car
+  # will be created so that only one injector attempts to create TLS certificates.
+  leaderElector:
+    enabled: true
+    image: "gcr.io/google_containers/leader-elector:0.4"
```
Comment (jasonodonnell) [inline values.yaml:36]: For consistency, can you make this:

```
image:
  repository: "gcr.io/google_containers/leader-elector"
  tag: "0.4"
```

File: templates/injector-leader-endpoint.yaml (hunk 2 above)
Comment (tomhjp) [reply templates/injector-leader-endpoint.yaml:4]: Good shout. Should I add the same to the secret too?

Comment (jasonodonnell) [reply templates/injector-leader-endpoint.yaml:4]: Yup, everything we make should have them.

Comment (tomhjp) [reply templates/injector-leader-endpoint.yaml:4]: Done

File: values.yaml (hunk 3 above)
Comment (tomhjp) [reply values.yaml]: Done

**Review by tvoran — APPROVED** [review]
I agree with Jason's comment about a helper function for the template line, and we may also want to add an election ttl configurable option: https://github.com/hashicorp/vault-k8s/blob/master/deploy/injector-deployment.yaml#L29

But I'd say neither of those are blockers, and can be done in subsequent PRs if we like.

**Comment by tomhjp** [comment]
I've added configurable TTL, and acceptance test is still passing locally:

```bash
bats -t test/acceptance/injector-leader-elector.bats
1..1
ok 1 injector: testing leader elector
```

**Review by tvoran — APPROVED** [review]

**Pull request merged**